### PR TITLE
Migrate Python tests to pytest

### PR DESCRIPTION
**Pull request by lgarrison**
From #258, we had an OpenMP error that only showed up in the Python tests. Therefore we decided to be a bit more organized about the Python tests, and migrate to pytest.  The code is pretty compact, and tests all ISA for all main API functions.

These tests all use the same data and configuration as the C tests, and hence we have the external, correct results to compare against.

Once this gets into master, we can merge/rebase master back into linearbinning. I have a corresponding `test_lin.py` that we can put into linearbinning once that happens.

### CODE REVIEW

**Comment by lgarrison** [comment]
@manodeep the Python 2.7 tests have trouble with importing this module structure. And looking ahead to `test_lin.py`, Python 2.7 some of the syntax there. Okay if I drop Python 2.7 from the Python test stage of the CI? Or even remove it from the CI matrix completely, unless we want to keep testing the build of the C extensions against Python 2.7.

**Comment by manodeep** [comment]
Only just got a chance to take a look - not sure I follow what's going on with the test failure - could it just be a missing `__init__.py` file?

We should try to support python 2.7 for at least Corrfunc v2.5, and then drop the py2.7 support from Corrfunc v2.6. Would it work to add a deprecation warning in v2.5?

**Comment by manodeep** [comment]
We should also test for some values of `nthreads` - say `[1, 2, 3, 4]` - perhaps with `isa=fastest`. That would (hopefully) detect any OpenMP threading issues

**Comment by lgarrison** [comment]
This version tests all ISA for the max number of threads (inferred from cores/affinity), and thread counts from 1 to `min(4,maxthreads)` for the fastest ISA.  Unfortunately, that's a lot of combos, so CI time has increased quite a bit; the Python tests alone take over 10 minutes.  I'm open to suggestions on smart ways to reduce the number of tests...

In addition, this does add one test case with `nthreads = ncores + 1`, just to check for correctness in that corner case.

**Comment by manodeep** [comment]
Looking at the CI failure, looks like `os.cpu_count()` does not exist in python2.7. That `AttributeError` should be fixable by a `try` but setting the actual number of threads correctly might be more difficult. Saw this [solution on SO](https://stackoverflow.com/a/1006301/2237582) - which we can rip off and keep only the POSIX relevant bits. What do you think?

**Comment by lgarrison** [comment]
Working now, I think!

**Comment by manodeep** [comment]
Still haven't had time to review - will look in the next 24-48 hours at the latest!

File: Corrfunc/tests/common.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    return ra, dec, cz, w
+
+
+def maxthreads():
+    '''Use as many threads as cores that are available to this process'''
+    import multiprocessing
+    
+    try:
+        maxthreads = len(os.sched_getaffinity(0))
+    except:
```
Comment (manodeep) [inline Corrfunc/tests/common.py:43]: Should this catch a specific exception?

File: Corrfunc/tests/common.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    return maxthreads
+
+
+def all_isa_nthreads(fastest_nthreads=4):
+    '''Test all ISA with maxthreads, and the fastest ISA with threads 1 to `fastest_nthreads`'''
+    mx = maxthreads()
+    # don't test with more threads than cores
+    fastest_nthreads = min(fastest_nthreads,mx)
+    # ... except for one test, which we'll force to have more threads than cores as a "stress test"
+    all_nthreads = list(range(1,fastest_nthreads+1)) + [mx+1,]
```
Comment (manodeep) [inline Corrfunc/tests/common.py:55]: This is PEP compatible, right? 

I would also suggest renaming fastest_nthreads to max_nthreads - which (I think) is what you are after. Additionally, line 55 is doing too much by itself - I am not sure I understand what's happening. Perhaps, split this into two lines to show that the normal usage about testing for all threads `[1, fastest_nthreads]` and then append to this list to show an additional `nthreads` test is done.

This function has a high cognitive load. From what I see this function is generating all allowed combinations of `ISAs` and `nthreads` (a list of `[1, max_nthreads]`) - therefore, the function name might be better off as `generate_isa_and_nthreads_combos` (in keeping with the convention of using a verb for a function name).

File: Corrfunc/tests/common.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    
+    combos = []
+    all_isa = ['fallback','sse42','avx','avx512f']
+    combos += [(isa,mx) for isa in all_isa]
+    combos += [('fastest',n) for n in all_nthreads]
+    
+    return combos
+
+
+def _check_against_reference(results, filename, results_cols=(-2, -1, 2), ref_cols=(0, 4, 1)):
```
Comment (manodeep) [inline Corrfunc/tests/common.py:65]: Given the naming choice for this function (that you call from a separate file) - perhaps remove the prefix `_`? Similarly, if a function is only being called within this file, then perhaps "reduce" the intended scope with a prefix `_` in the name?

File: Corrfunc/tests/common.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    # results is output of Python function
+    # filename the reference counts
+    # ref_cols in order of npairs, weightavg, rpavg
+    refs = np.loadtxt(filename, unpack=True, usecols=ref_cols)
+    for ii, items in enumerate(results):
+        for icol, ref in zip(results_cols, refs):
```
Comment (manodeep) [inline Corrfunc/tests/common.py:71]: I am afraid of hardcoding column numbers - is there any other way around that?

File: Corrfunc/tests/common.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            assert np.allclose(items[icol], ref[ii])
```
Comment (manodeep) [inline Corrfunc/tests/common.py:72]: Is the convention to fail during tests - (ie within the developer's realm) or is there an expectation for a helpful message associated with why the test failed (+ any other relevant contextual info)?

Comment (manodeep) [reply Corrfunc/tests/common.py:72]: And do is it generally recommended to use the default `np.allclose` for tests? Is that what we do in the C tests?

**Review by manodeep — COMMENTED** [review]
I did a quick skim of the code - looks pretty good! 

Do we need to figure out a way to reduce the runtime?

File: Corrfunc/tests/common.py (hunk 2 above)
Comment (lgarrison) [reply Corrfunc/tests/common.py:55]: > I would also suggest renaming fastest_nthreads to max_nthreads - which (I think) is what you are after
`fastest_nthreads` is the number of threads that we will use in the thread sweep, which only uses the fastest ISA. This will likely not be the same as the number of threads used in the ISA sweep! Agreed this is a subtle point, let me refactor this...

File: Corrfunc/tests/common.py (hunk 4 above)
Comment (lgarrison) [reply Corrfunc/tests/common.py:71]: The underlying files don't have column names, so any solution is going to involve a hard-coded mapping of column numbers. Not pretty, but these files are static and the C code access the columns in the same way (hard-coded column numbers), so I think it's pretty safe.

Comment (lgarrison) [reply Corrfunc/tests/common.py:71]: Having said that, I couldn't help myself, and I did end up rewriting this function in a way that uses column names! There's one fewer hardcoded mapping since the Python results struct is now accessed by column name instead of number.

**Comment by lgarrison** [comment]
Here's a version that refactors the reference check to use column names where possible! And, most importantly, the `npairs` check now uses an integer equality test rather than some tolerance.

On the runtime, I'm considering the fact that the ISA sweep already tests the multi-threaded correctness. So what about reducing the thread sweep to two values: `1` and `ncore+1`. This tests single-threaded and "oversubscribed" correctness.  Because the ISA sweep uses `ncores` threads, `ncores+1` also ensures we test both an even and odd number of threads.  I think that's pretty good coverage: single-threaded, even-number multithreaded, odd-number multithreaded, and oversubscribed multithreaded.

File: Corrfunc/io.py [hunk 6]
```diff
@@ -19,7 +19,7 @@
 __all__ = ('read_fastfood_catalog', 'read_ascii_catalog', 'read_catalog')
 
 
-def read_fastfood_catalog(filename, return_dtype=None, need_header=None):
+def read_fastfood_catalog(filename, return_dtype=None, need_weights=None, need_header=None):
```
Comment (manodeep) [inline Corrfunc/io.py:22]: I don't see `need_header` being used anywhere in the code-base - should we remove this keyword?

File: Corrfunc/tests/common.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    
+    # this is the max number of threads we will use when testing multiple values of nthreads
+    # we will cap this at the number of cores
+    thread_sweep_max = min(thread_sweep_max,mx)
+    all_nthreads = list(range(1,thread_sweep_max+1))
+    # ... except for one test, which we'll force to have more threads than cores as a "stress test"
+    all_nthreads += [mx+1]
+    
+    combos = []
+    all_isa = ['fallback','sse42','avx','avx512f']
```
Comment (manodeep) [inline Corrfunc/tests/common.py:61]: `wp` also has an `AVX2` kernel - should we add that in specifically somehow - or just ignore that from python?

File: Corrfunc/tests/test_mocks.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+import numpy as np
+
+from Corrfunc.tests.common import Mr19_mock_northonly, Mr19_randoms_northonly
+from Corrfunc.tests.common import (check_against_reference,
+                                   check_vpf_against_reference)
+from Corrfunc.tests.common import generate_isa_and_nthreads_combos
+    
+
+@pytest.mark.parametrize('isa,nthreads', generate_isa_and_nthreads_combos())
+def test_DDrppi_mocks(Mr19_mock_northonly, isa, nthreads):
```
Comment (manodeep) [inline Corrfunc/tests/test_mocks.py:15]: Whoa - is this the normal convention? `Mr19_mock_northonly` is both a function and a parameter? Particularly, on line 24, is there a function call involved or is `pytests` caching that function call return value?

File: Corrfunc/tests/test_theory.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+                     "../../theory/tests/", "bins")
+    autocorr = 1
+    periodic = 1
+    
+    x,y,z,w = gals_Mr19
+    results_DD = DD(autocorr, nthreads, binfile, x, y, z,
+                            weights1=w, weight_type='pair_product',
+                            periodic=periodic, boxsize=boxsize,
+                            output_ravg=True, verbose=True,
+                            isa=isa)
```
Comment (manodeep) [inline Corrfunc/tests/test_theory.py:29]: Whitespace issues here - the second line of parameters should probably align with the opening bracket on the first line (or whatever PEP recommends). Similar alignment are there elsewhere - will you please fix if possible. 

(We really need to get a linter into our CI/pre-commit)

**Review by manodeep — APPROVED** [review]

File: Corrfunc/tests/common.py (hunk 4 above)
Comment (manodeep) [reply Corrfunc/tests/common.py]: Hehehehee - I know that feeling :)

**Comment by manodeep** [comment]
Everything looks good to me - I made some minor comments, mostly related to whitespace and (likely) pytests wizardry

**Comment by lgarrison** [comment]
Should I go ahead and reduce the test cases as proposed above?

File: Corrfunc/tests/test_mocks.py (hunk 8 above)
Comment (lgarrison) [reply Corrfunc/tests/test_mocks.py:15]: Yes! It's called a fixture: https://docs.pytest.org/en/6.2.x/fixture.html#back-to-fixtures

File: Corrfunc/tests/test_theory.py (hunk 9 above)
Comment (lgarrison) [reply Corrfunc/tests/test_theory.py:29]: Yeah, there's a lot of whitespace issues! I clearly need to upgrade to an editor that will flag PEP8 issues on-the-fly. But I think the codebase is littered with whitespace problems, and it's probably more efficient if we do a big PEP8 linting pass on the entire codebase, rather than playing whack-a-mole each time we open a PR. So I may leave this for now, if that's okay.

**Comment by manodeep** [comment]
> Should I go ahead and reduce the thread test cases as proposed above?

Ahh yes - that should be a good enough nthreads coverage. And as long as the single thread agrees with a few different multiple-thread, we are probably okay (haven't tested that though)

**Comment by lgarrison** [comment]
Great -- single- & multi-threaded both agree with the known answer, so this is a pretty good test I think.

I'll merge this and then create a `test_lin.py` in `linearbinning`, and once that passes we'll be ready to merge `linearbinning`!
